MuvrKit/MKWatchConnectivity.swift

```diff
@@ -88,28 +108,233 @@ public class MKConnectivity : NSObject, WCSessionDelegate {
-            WCSession.defaultSession().transferUserInfo(message)
+    public func endLastSession() {
+        if let (session, props) = mostImportantSessionsEntry() {
+            sessions[session] = props.with(end: NSDate())
+            execute()
         }
-        currentSession = nil
     }
+    
+    ///
+    /// Returns the most important session for processing, if available
+    ///
+    private func mostImportantSessionsEntry() -> (MKExerciseSession, MKExerciseSessionProperties)? {
+        // pick the not-yet-ended session first
+        for (session, props) in sessions {
+            if props.end == nil {
+                return (session, props)
+            }
+        }
         
+        // then whichever one remains
+        return sessions.first
+    }
+    
     ///
-    /// Starts the exercise session with the given 
+    /// Implements the protocol for the W -> P communication by collecting the data from the sensor recorder,
+    /// constructing the messages and dealing with session clean-up.
     ///
-    public func startSession(exerciseModelMetadata exerciseModelMetadata: MKExerciseModelMetadata, demo: Bool) {
-        if currentSession != nil { endSession() }
+    public func execute() {
+        func getSamples(from from: NSDate, to: NSDate, requireAll: Bool, demo: Bool) -> MKSensorData? {
+            var simulatedSamples = demo
+            
+            #if (arch(i386) || arch(x86_64))
+                simulatedSamples = true
+            #endif
+            
+            let duration = to.timeIntervalSinceDate(from)
+            let sampleCount = 3 * 50 * Int(duration)
+            
+            if simulatedSamples {
+                let samples = (0..<sampleCount).map { _ in return Float(0) }
+                return try! MKSensorData(types: [.Accelerometer(location: .LeftWrist)], start: 0, samplesPerSecond: 50, samples: samples)
+            } else {
+                return recorder.accelerometerDataFromDate(from, toDate: to).flatMap { (recordedData: CMSensorDataList) -> MKSensorData? in
+                    let samples = recordedData.enumerate().flatMap { (_, e) -> [Float] in
+                        if let data = e as? CMRecordedAccelerometerData {
+                            return [Float(data.acceleration.x), Float(data.acceleration.y), Float(data.acceleration.z)]
+                        }
+                        return []
+                    }
+                    // remember to check for truly complete block
+                    // it's OK to leave out the very last window
+                    if samples.count < (sampleCount - 1200) && requireAll {
+                        NSLog("Not yet flushed buffer. Expected \(sampleCount), got \(samples.count)")
+                        return nil
+                    }
+                    return try! MKSensorData(types: [.Accelerometer(location: .LeftWrist)], start: 0, samplesPerSecond: 50, samples: samples)
```

Comment:
I don't really like it either, but the motivation is that the ``MKSensorData`` can pad small gaps. The padding only kicks in when ``start > 0``.

It will be very useful when we have multiple sensors.